router/failed-events-manager.go

```diff
@@ -0,0 +1,177 @@
+			if err != nil {
+				panic(err)
+			}
+		}
+
+		stmt.Close()
+	}
+}
+
+func (fem *FailedEventsManagerT) DropFailedRecordIDs(taskRunID string) {
+	if !failedKeysEnabled {
+		return
+	}
+	//Drop table
+	table := fmt.Sprintf(`%s_%s`, failedKeysTablePrefix, taskRunID)
+	sqlStatement := fmt.Sprintf(`DROP TABLE IF EXISTS %s`, table)
+	_, err := fem.dbHandle.Exec(sqlStatement)
+	if err != nil {
+		pkgLogger.Errorf("Failed to drop table %s with error: %v", taskRunID, err)
+	}
+}
+
+func (fem *FailedEventsManagerT) FetchFailedRecordIDs(taskRunID string) []*FailedEventRowT {
+	if !failedKeysEnabled {
+		return []*FailedEventRowT{}
+	}
+	failedEvents := make([]*FailedEventRowT, 0)
+
+	var rows *sql.Rows
+	var err error
+	table := fmt.Sprintf(`%s_%s`, failedKeysTablePrefix, taskRunID)
+	sqlStatement := fmt.Sprintf(`SELECT %[1]s.destination_id, %[1]s.record_id
+                                             FROM %[1]s `, table)
+	rows, err = fem.dbHandle.Query(sqlStatement)
+	if err != nil {
+		pkgLogger.Errorf("Failed to fetch from table %s with error: %v", taskRunID, err)
+		return failedEvents
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var failedEvent FailedEventRowT
+		err := rows.Scan(&failedEvent.DestinationID, &failedEvent.RecordID)
+		if err != nil {
+			panic(err)
+		}
+		failedEvents = append(failedEvents, &failedEvent)
+	}
+
+	return failedEvents
+}
+
+func CleanFailedRecordsTableProcess() {
+	if !failedKeysEnabled {
+		return
+	}
+	for {
+		dbHandle, err := sql.Open("postgres", jobsdb.GetConnectionString())
+		if err != nil {
+			panic(err)
+		}
+		failedKeysLike := failedKeysTablePrefix + "%"
+		failedKeysTableQuery := fmt.Sprintf(`SELECT table_name
+													FROM information_schema.tables
+													WHERE table_schema='public' AND table_type='BASE TABLE' AND table_name ilike '%s'`, failedKeysLike)
+		rows, err := dbHandle.Query(failedKeysTableQuery)
+		if err != nil {
+			panic(err)
+		}
+		for rows.Next() {
+			var table string
+			err = rows.Scan(&table)
+			if err != nil {
+				pkgLogger.Errorf("Failed to scan failed keys table %s with error: %v", table, err)
+				return
+			}
+			latestCreatedAtQuery := fmt.Sprintf(`SELECT created_at from %s order by created_at desc limit 1`, table)
+			row := dbHandle.QueryRow(latestCreatedAtQuery)
+			var latestCreatedAt time.Time
+			err = row.Scan(&latestCreatedAt)
+			if err != nil && err != sql.ErrNoRows {
+				pkgLogger.Errorf("Failed to fetch records from failed keys table %s with error: %v", table, err)
+				continue
+			}
+			currentTime := time.Now()
+			diff := currentTime.Sub(latestCreatedAt)
+			if diff > failedKeysExpire {
+				dropQuery := fmt.Sprintf(`DROP TABLE IF EXISTS %s`, table)
+				rows, err = dbHandle.Query(dropQuery)
+				if err != nil {
+					pkgLogger.Errorf("Failed to drop table %s with error: %v", table, err)
+				}
+			}
+		}
+		dbHandle.Close()
+		time.Sleep(failedKeysCleanUpSleep)
```

Comment:
minor: Not sure whats a better option. An ever running process with sleep vs using a scheduled timer with registered function.

Reply:
Added a task in backlog
https://www.notion.so/rudderstacks/cfe8e8ef60a34485a373e789b7b6933a?v=a07a0b1350f84d46af9792260b39a194&p=a91b5dcdfe414780bb7bd7648c9d4c05